src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,609 @@
+            }
+        }
+
+        // legacy
+        public JSObject? GetCSOwnedObjectByJSHandle(nint jsHandle, int shouldAddInflight)
+        {
+            lock (this)
+            {
+                if (ThreadCsOwnedObjects.TryGetValue(jsHandle, out WeakReference<JSObject>? reference))
+                {
+                    reference.TryGetTarget(out JSObject? jsObject);
+                    if (shouldAddInflight != 0)
+                    {
+                        jsObject?.AddInFlight();
+                    }
+                    return jsObject;
+                }
+            }
+            return null;
+        }
+
+        // legacy
+        public JSObject CreateCSOwnedProxy(nint jsHandle, LegacyHostImplementation.MappedType mappedType, int shouldAddInflight)
+        {
+            lock (this)
+            {
+                JSObject? res = null;
+                if (!ThreadCsOwnedObjects.TryGetValue(jsHandle, out WeakReference<JSObject>? reference) ||
+                !reference.TryGetTarget(out res) ||
+                res.IsDisposed)
+                {
+#pragma warning disable CS0612 // Type or member is obsolete
+                    res = mappedType switch
+                    {
+                        LegacyHostImplementation.MappedType.JSObject => new JSObject(jsHandle, JSProxyContext.MainThreadContext),
+                        LegacyHostImplementation.MappedType.Array => new Array(jsHandle),
+                        LegacyHostImplementation.MappedType.ArrayBuffer => new ArrayBuffer(jsHandle),
+                        LegacyHostImplementation.MappedType.DataView => new DataView(jsHandle),
+                        LegacyHostImplementation.MappedType.Function => new Function(jsHandle),
+                        LegacyHostImplementation.MappedType.Uint8Array => new Uint8Array(jsHandle),
+                        _ => throw new ArgumentOutOfRangeException(nameof(mappedType))
+                    };
+#pragma warning restore CS0612 // Type or member is obsolete
+                    ThreadCsOwnedObjects[jsHandle] = new WeakReference<JSObject>(res, trackResurrection: true);
+                }
+                if (shouldAddInflight != 0)
+                {
+                    res.AddInFlight();
+                }
+                return res;
+            }
+        }
+
+        #endregion
+
+        #region Dispose
+
+        private void Dispose(bool disposing)
+        {
+            lock (this)
+            {
+                if (!_isDisposed)
+                {
+#if FEATURE_WASM_THREADS
+                    if (!IsCurrentThread()) throw new InvalidOperationException("JSProxyContext must be disposed on the thread which owns it.");
+                    AssertOperationStack(0);
+                    _OperationStack = null;
+#endif
+
+                    // TODO: free unmanaged resources (unmanaged objects) and override finalizer
+                    foreach (var jsObjectWeak in ThreadCsOwnedObjects.Values)
```

Comment:
Could the `jso.Dispose` mutate `ThreadCsOwnedObjects`? If so, this would crash and we need to duplicate the `Values` list before iterating it (`ToList` should be sufficient)

Reply:
good catch, made copy